truss/templates/server/common/schema.py

```diff
@@ -0,0 +1,164 @@
+import inspect
+from types import MappingProxyType
+from typing import (
+    Any,
+    AsyncGenerator,
+    Awaitable,
+    Generator,
+    Optional,
+    Tuple,
+    Union,
+    get_args,
+    get_origin,
+)
+
+from pydantic import BaseModel
+
+
+class TrussSchema(BaseModel):
+    input_type: type
+    output_type: Optional[type]
+    supports_streaming: bool
+
+    @classmethod
+    def from_signature(
+        cls, input_parameters: MappingProxyType, output_annotation: Any
+    ) -> Optional["TrussSchema"]:
+        """
+        Create a TrussSchema from a function signature if annotated, else returns None
+        """
+
+        input_type = _parse_input_type(input_parameters)
+        ouput_type = _parse_output_type(output_annotation)
+
+        if not input_type or not ouput_type:
+            return None
+
+        output_type, supports_streaming = ouput_type
+
+        return cls(
+            input_type=input_type,
+            output_type=output_type,
+            supports_streaming=supports_streaming,
+        )
+
+
+def _parse_input_type(input_parameters: MappingProxyType) -> Optional[type]:
+    parameter_types = list(input_parameters.values())
+
+    if len(parameter_types) > 1:
+        return None
+
+    input_type = parameter_types[0].annotation
+
+    if (
+        input_type == inspect.Signature.empty
+        or not isinstance(input_type, type)
+        or not issubclass(input_type, BaseModel)
+    ):
+        return None
+
+    return input_type
+
+
+def _parse_output_type(output_annotation: Any) -> Optional[Tuple[Optional[type], bool]]:
+    """
+    Therea are 4 possible cases for output_annotation:
+    1. Data object -- represented by a Pydantic BaseModel
+    2. Streaming -- represented by a Generator or AsyncGenerator
+    3. Async -- represented by an Awaitable
+    4. Streaming or data object -- it is possible for a function to return both
+       a data object and a streaming object. This is represented by a Union of
+       a Pydantic BaseModel and a Generator or AsyncGenerator
+
+    If the output_annotation does not match one of these cases, returns None
+    """
+    if isinstance(output_annotation, type) and issubclass(output_annotation, BaseModel):
+        return output_annotation, False
+
+    if _is_generator_type(output_annotation):
+        return None, True
```

Comment:
Can a streaming model's return type be annotated `Generator[ModelOutput, None, None]`? If so, should we try to parse the output type instead of returning None (similar to what we do for awaitables)?

Reply:
`Generator[ModelOutput...]` would not be a valid schema. You can only stream back strings